// coddyger/dist/config/swagger.config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.swaggerUiConfig = exports.swaggerConfig = void 0;
const globals_1 = __importDefault(require("../globals"));
exports.swaggerConfig = {
    info: {
        title: globals_1.default.appName,
        description: globals_1.default.description,
        version: globals_1.default.server.apiVersion,
    },
    externalDocs: {
        url: "https://swagger.io",
        description: "Find more info here",
    },
    // host: "localhost:8080",
    schemes: ["http"],
    consumes: ["application/json"],
    produces: ["application/json"],
    tags: [
    //   { name: "user", description: "User related end-points" },
    //   { name: "code", description: "Code related end-points" },
    ],
    definitions: {},
    securityDefinitions: {
        apiKey: {
            type: "apiKey",
            name: "apiKey",
            in: "header",
        },
    },
};
exports.swaggerUiConfig = {
    routePrefix: "/docs",
    uiConfig: {
        docExpansion: "none",
        deepLinking: false,
    },
    uiHooks: {
        onRequest: function (request, reply, next) {
            next();
        },
        preHandler: function (request, reply, next) {
            next();
        },
    },
    staticCSP: true,
    transformStaticCSP: (header) => header,
    transformSpecification: (swaggerObject, request, reply) => {
        return swaggerObject;
    },
    transformSpecificationClone: true,
};
