// coddyger/src/config/swagger.config.ts
import env from "../globals";

export const swaggerConfig: any = {
  info: {
    title: env.appName,
    description: env.description,
    version: env.server.apiVersion,
  },
  externalDocs: {
    url: "https://swagger.io",
    description: "Find more info here",
  },
  // host: "localhost:8080",
  schemes: ["http"],
  consumes: ["application/json"],
  produces: ["application/json"],
  tags: [
    //   { name: "user", description: "User related end-points" },
    //   { name: "code", description: "Code related end-points" },
  ],
  definitions: {},
  securityDefinitions: {
    apiKey: {
      type: "apiKey",
      name: "apiKey",
      in: "header",
    },
  },
};
export const swaggerUiConfig: any = {
  routePrefix: "/docs",
  uiConfig: {
    docExpansion: "none",
    deepLinking: false,
  },
  uiHooks: {
    onRequest: function (request, reply, next) {
      next();
    },
    preHandler: function (request, reply, next) {
      next();
    },
  },
  staticCSP: true,
  transformStaticCSP: (header) => header,
  transformSpecification: (swaggerObject, request, reply) => {
    return swaggerObject;
  },
  transformSpecificationClone: true,
};
