// coddyger/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Controller = exports.SequelizeDataAccess = exports.MongooseDataAccess = exports.MongoDbDao = exports.getDatabaseAccess = exports.LogLevel = exports.defines = exports.env = exports.AxiosService = exports.TransporterService = exports.LoggerService = void 0;
const coddyger_1 = __importDefault(require("./coddyger"));
const globals_1 = __importStar(require("./globals"));
exports.env = globals_1.default;
Object.defineProperty(exports, "defines", { enumerable: true, get: function () { return globals_1.defines; } });
const services_1 = require("./services");
Object.defineProperty(exports, "AxiosService", { enumerable: true, get: function () { return services_1.AxiosService; } });
Object.defineProperty(exports, "LogLevel", { enumerable: true, get: function () { return services_1.LogLevel; } });
Object.defineProperty(exports, "LoggerService", { enumerable: true, get: function () { return services_1.LoggerService; } });
Object.defineProperty(exports, "TransporterService", { enumerable: true, get: function () { return services_1.TransporterService; } });
const database_1 = require("./database");
Object.defineProperty(exports, "getDatabaseAccess", { enumerable: true, get: function () { return database_1.getDatabaseAccess; } });
Object.defineProperty(exports, "MongoDbDao", { enumerable: true, get: function () { return database_1.MongoDbDao; } });
Object.defineProperty(exports, "MongooseDataAccess", { enumerable: true, get: function () { return database_1.MongooseDataAccess; } });
Object.defineProperty(exports, "SequelizeDataAccess", { enumerable: true, get: function () { return database_1.SequelizeDataAccess; } });
const controller_1 = __importDefault(require("./controllers/controller"));
exports.Controller = controller_1.default;
exports.default = coddyger_1.default;
