// coddyger/dist/services/module.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModuleService = void 0;
class ModuleService {
    add() {
        this.plop.setGenerator('module', {
            description: 'Générer un module',
            prompts: [{
                    type: 'input',
                    name: 'name',
                    message: 'Nom du module:',
                }],
            actions: [
                {
                    type: 'add',
                    path: '../modules/{{name}}/index.ts',
                    templateFile: 'templates/no-user/index.hbs',
                },
                {
                    type: 'add',
                    path: '../modules/{{name}}/{{name}}.model.ts',
                    templateFile: 'templates/no-user/model.hbs',
                },
                {
                    type: 'add',
                    path: '../modules/{{name}}/{{name}}.controller.ts',
                    templateFile: 'templates/no-user/controller.hbs',
                },
                {
                    type: 'add',
                    path: '../modules/{{name}}/{{name}}.service.ts',
                    templateFile: 'templates/no-user/service.hbs',
                },
                {
                    type: 'add',
                    path: './routes/{{name}}.route.ts',
                    templateFile: 'templates/no-user/route.hbs',
                },
            ],
        });
    }
}
exports.ModuleService = ModuleService;
